library/std/src/io/buffered/bufreader.rs

```diff
@@ -271,6 +271,35 @@ impl<R: Read> Read for BufReader<R> {
         Ok(nread)
     }
 
+    // Small read_exacts from a BufReader are extremely common when used with a deserializer.
+    // The default implementation calls read in a loop, which results in surprisingly poor code
+    // generation for the common path where the buffer has enough bytes to fill the passed-in
+    // buffer.
+    fn read_exact(&mut self, mut buf: &mut [u8]) -> io::Result<()> {
+        if self.buffer().len() >= buf.len() {
+            buf.copy_from_slice(&self.buffer()[..buf.len()]);
+            self.consume(buf.len());
+            return Ok(());
+        }
+
+        while !buf.is_empty() {
```

Comment:
We should be able to introduce a `read_exact` function with this default implementation and call it here and in `Read::read_exact` so that it’s clear this is our fallback.